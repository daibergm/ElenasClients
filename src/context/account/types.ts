export enum ActionTypes {
  LOGIN_ATTEMPT = '@ACCOUNT_LOGIN_ATTEMPT',
  LOGIN_SUCCESS = '@ACCOUNT_LOGIN_SUCCESS',
  LOGIN_FAILURE = '@ACCOUNT_LOGIN_FAILURE',
  LOGOUT = '@ACCOUNT_LOGOUT',
}

export type Action = {
  type:
    | ActionTypes.LOGIN_ATTEMPT
    | ActionTypes.LOGIN_SUCCESS
    | ActionTypes.LOGIN_FAILURE
    | ActionTypes.LOGOUT;
  payload?: string;
};
