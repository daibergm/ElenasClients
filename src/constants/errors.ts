interface IErrors<T> {
  [id: string]: T;
}

export const ERRORS: IErrors<string> = {
  ValidationErrors: 'Celular o contraseña incorrectos.',
  ClientValidationErrors: 'Falta  información o el ciente ya existe.',
  LoginFailure: 'Error desconocido, intente de nuevo mas tarde.',
};
