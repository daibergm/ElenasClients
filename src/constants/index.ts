export * from './routes';
export * from './testIds';
export * from './errors';

export const STORAGE_SESSION_KEY = '@ACCOUNT_SESSION';
