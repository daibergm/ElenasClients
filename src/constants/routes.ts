export const LOGIN_ROUTE = 'LOGIN';
export const CLIENTS_ROUTE = 'CLIENTS';
export const CLIENT_ROUTE = 'CLIENT';
