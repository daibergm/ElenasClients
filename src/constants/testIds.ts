export const TEST_IDS = {
  submitButton: 'submitButton',
  cellphoneInput: 'cellphoneInput',
  passwordInput: 'passwordInput',
};
