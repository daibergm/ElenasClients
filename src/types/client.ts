export type Client = {
  id: number;
  firstName: string;
  lastName: string;
  cellphone: string;
};
