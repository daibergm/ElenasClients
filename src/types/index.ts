export * from './routes';
export * from './user';
export * from './client';
