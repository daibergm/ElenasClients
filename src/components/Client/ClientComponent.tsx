import React from 'react';
import { View, ScrollView, Text } from 'react-native';
import { Formik } from 'formik';
import * as yup from 'yup';
import { isEmpty } from 'lodash';

// @Assets
import styles from './styles';

// @Types
import { Client } from '../../types/';

// @Components
import { Button, Input } from '../common';

// @Constants
import { TEST_IDS, ERRORS } from '../../constants/';

const validationSchema = yup.object().shape({
  firstName: yup.string().trim().required('Este campo es obligatorio.'),
  lastName: yup.string().trim().required('Este campo es obligatorio.'),
  cellphone: yup.string().trim().required('Este campo es obligatorio.'),
  address: yup.string().trim().required('Este campo es obligatorio.'),
});

interface Props {
  submitFunction: (values: Client) => void;
  loading: boolean;
  mutationLoading: boolean;
  apiError?: string;
  id?: number;
  client: Client;
}

function ClientComponent({
  submitFunction,
  loading,
  mutationLoading,
  id,
  client,
  apiError,
}: Props) {
  const onSubmit = (values: Client) => submitFunction(values);

  return (
    <View style={styles.container}>
      <Formik
        initialValues={client}
        validationSchema={validationSchema}
        onSubmit={onSubmit}
        enableReinitialize>
        {({
          values,
          handleChange,
          errors,
          setFieldTouched,
          isValid,
          handleSubmit,
          touched,
          dirty,
        }) => (
          <ScrollView
            alwaysBounceVertical={false}
            showsVerticalScrollIndicator={false}
            style={styles.scroll}
            keyboardShouldPersistTaps="handled">
            <View style={styles.subContainer}>
              <Input
                testID={TEST_IDS.firstNameInput}
                value={values.firstName}
                placeholder="Nombre"
                errorMessage={
                  touched.firstName ? (errors.firstName as string) : undefined
                }
                onChangeText={handleChange('firstName')}
                onBlur={() => setFieldTouched('firstName')}
              />
              <Input
                testID={TEST_IDS.lastNameInput}
                value={values.lastName}
                placeholder="Apellido"
                errorMessage={
                  touched.lastName ? (errors.lastName as string) : undefined
                }
                onChangeText={handleChange('lastName')}
                onBlur={() => setFieldTouched('lastName')}
              />
              <Input
                testID={TEST_IDS.cedulaInput}
                value={values.cedula}
                placeholder="Cédula"
                errorMessage={
                  touched.cedula ? (errors.cedula as string) : undefined
                }
                onChangeText={handleChange('cedula')}
                onBlur={() => setFieldTouched('cedula')}
              />
              <Input
                testID={TEST_IDS.emailInput}
                value={values.email}
                placeholder="Correo electrónico"
                errorMessage={
                  touched.email ? (errors.email as string) : undefined
                }
                onChangeText={handleChange('email')}
                onBlur={() => setFieldTouched('email')}
              />
              <Input
                testID={TEST_IDS.cellphoneInput}
                value={values.cellphone}
                placeholder="Celular"
                errorMessage={
                  touched.cellphone ? (errors.cellphone as string) : undefined
                }
                onChangeText={handleChange('cellphone')}
                onBlur={() => setFieldTouched('cellphone')}
              />
              <Input
                testID={TEST_IDS.addressInput}
                value={values.address}
                placeholder="Dirección"
                errorMessage={
                  touched.address ? (errors.address as string) : undefined
                }
                onChangeText={handleChange('address')}
                onBlur={() => setFieldTouched('address')}
              />
              <View style={styles.buttonContainer}>
                <Button
                  title={id ? 'EDITAR' : 'CREAR'}
                  testID={TEST_IDS.submitButton}
                  disabled={(!isValid || !dirty) && !id}
                  onPress={handleSubmit}
                  loading={loading || mutationLoading}
                />
              </View>
              {!isEmpty(apiError) && (
                <View>
                  <Text style={styles.errorLabel}>
                    {ERRORS[apiError as string]}
                  </Text>
                </View>
              )}
            </View>
          </ScrollView>
        )}
      </Formik>
    </View>
  );
}

export default ClientComponent;
