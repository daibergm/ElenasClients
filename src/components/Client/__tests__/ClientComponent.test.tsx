import React from 'react';
import { render, fireEvent, waitFor } from '@testing-library/react-native';

// @Components
import ClientComponent from '../ClientComponent';

// @Constants
import { TEST_IDS } from '../../../constants/';

// @Types
import { Client } from '../../../types';

// @Vars
const initialValues: Client = {
  firstName: '',
  lastName: '',
  cedula: '',
  email: '',
  cellphone: '',
  address: '',
};

describe('Client test suits', () => {
  const props = {
    loading: false,
    mutationLoading: false,
    submitFunction: () => jest.fn(),
    client: initialValues,
  };

  it('Render correctly', () => {
    const client = render(<ClientComponent {...props} />);
    expect(client).toMatchSnapshot();
  });
  it('Submit function', () => {
    const { getByTestId } = render(<ClientComponent {...props} />);
    const submitButton = getByTestId(TEST_IDS.submitButton);
    const firstNameInput = getByTestId(TEST_IDS.firstNameInput);
    const lastNameInput = getByTestId(TEST_IDS.lastNameInput);
    const cellphoneInput = getByTestId(TEST_IDS.cellphoneInput);
    const addressInput = getByTestId(TEST_IDS.addressInput);
    fireEvent.changeText(firstNameInput, 'John');
    fireEvent.changeText(lastNameInput, 'Doe');
    fireEvent.changeText(cellphoneInput, '1234567890');
    fireEvent.changeText(addressInput, 'Calle 10');
    fireEvent.press(submitButton);
    waitFor(() => {
      expect(props.submitFunction).toHaveBeenCalled();
    });
  });
  it('Loading button', () => {
    const { getByTestId } = render(
      <ClientComponent {...props} loading mutationLoading />,
    );
    const submitButton = getByTestId(TEST_IDS.submitButton);
    expect(submitButton.props.accessibilityState).toStrictEqual({
      disabled: true,
      busy: true,
    });
  });
  it('Edit form', () => {
    const { getByTestId } = render(<ClientComponent {...props} id={10} />);
    const submitButton = getByTestId(TEST_IDS.editButton);
    expect(submitButton).toBeTruthy();
  });
  it('Error label', () => {
    const { getByTestId } = render(
      <ClientComponent {...props} apiError="Test error" />,
    );
    const errorLabel = getByTestId(TEST_IDS.errorLabel);
    expect(errorLabel).toBeTruthy();
  });
});
