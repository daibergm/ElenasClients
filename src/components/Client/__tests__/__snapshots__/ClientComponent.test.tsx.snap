// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Client test suits Render correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "flexDirection": "row",
    }
  }
>
  <RCTScrollView
    alwaysBounceVertical={false}
    keyboardShouldPersistTaps="handled"
    showsVerticalScrollIndicator={false}
    style={
      Object {
        "paddingHorizontal": 15,
        "width": "100%",
      }
    }
  >
    <View>
      <View
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <View
          style={
            Object {
              "paddingHorizontal": 10,
              "width": "100%",
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "borderBottomColor": "rgba(0, 173, 238, 0.65)",
                "borderBottomWidth": 1,
                "borderColor": "#86939e",
                "flexDirection": "row",
                "height": 48,
                "transform": Array [
                  Object {
                    "translateX": 0,
                  },
                ],
              }
            }
          >
            <TextInput
              editable={true}
              onBlur={[Function]}
              onChangeText={[Function]}
              placeholder="Nombre"
              placeholderTextColor="rgba(0, 173, 238, 0.65)"
              style={
                Object {
                  "color": "#00adee",
                  "flex": 1,
                  "fontSize": 16,
                  "minHeight": 40,
                }
              }
              testID="firstNameInput"
              underlineColorAndroid="transparent"
              value=""
            />
          </View>
          <Text
            style={
              Object {
                "color": "#dc3545",
                "fontSize": 12,
                "margin": 5,
              }
            }
          />
        </View>
        <View
          style={
            Object {
              "paddingHorizontal": 10,
              "width": "100%",
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "borderBottomColor": "rgba(0, 173, 238, 0.65)",
                "borderBottomWidth": 1,
                "borderColor": "#86939e",
                "flexDirection": "row",
                "height": 48,
                "transform": Array [
                  Object {
                    "translateX": 0,
                  },
                ],
              }
            }
          >
            <TextInput
              editable={true}
              onBlur={[Function]}
              onChangeText={[Function]}
              placeholder="Apellido"
              placeholderTextColor="rgba(0, 173, 238, 0.65)"
              style={
                Object {
                  "color": "#00adee",
                  "flex": 1,
                  "fontSize": 16,
                  "minHeight": 40,
                }
              }
              testID="lastNameInput"
              underlineColorAndroid="transparent"
              value=""
            />
          </View>
          <Text
            style={
              Object {
                "color": "#dc3545",
                "fontSize": 12,
                "margin": 5,
              }
            }
          />
        </View>
        <View
          style={
            Object {
              "paddingHorizontal": 10,
              "width": "100%",
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "borderBottomColor": "rgba(0, 173, 238, 0.65)",
                "borderBottomWidth": 1,
                "borderColor": "#86939e",
                "flexDirection": "row",
                "height": 48,
                "transform": Array [
                  Object {
                    "translateX": 0,
                  },
                ],
              }
            }
          >
            <TextInput
              editable={true}
              onBlur={[Function]}
              onChangeText={[Function]}
              placeholder="Cédula"
              placeholderTextColor="rgba(0, 173, 238, 0.65)"
              style={
                Object {
                  "color": "#00adee",
                  "flex": 1,
                  "fontSize": 16,
                  "minHeight": 40,
                }
              }
              testID="cedulaInput"
              underlineColorAndroid="transparent"
              value=""
            />
          </View>
          <Text
            style={
              Object {
                "color": "#dc3545",
                "fontSize": 12,
                "margin": 5,
              }
            }
          />
        </View>
        <View
          style={
            Object {
              "paddingHorizontal": 10,
              "width": "100%",
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "borderBottomColor": "rgba(0, 173, 238, 0.65)",
                "borderBottomWidth": 1,
                "borderColor": "#86939e",
                "flexDirection": "row",
                "height": 48,
                "transform": Array [
                  Object {
                    "translateX": 0,
                  },
                ],
              }
            }
          >
            <TextInput
              editable={true}
              onBlur={[Function]}
              onChangeText={[Function]}
              placeholder="Correo electrónico"
              placeholderTextColor="rgba(0, 173, 238, 0.65)"
              style={
                Object {
                  "color": "#00adee",
                  "flex": 1,
                  "fontSize": 16,
                  "minHeight": 40,
                }
              }
              testID="emailInput"
              underlineColorAndroid="transparent"
              value=""
            />
          </View>
          <Text
            style={
              Object {
                "color": "#dc3545",
                "fontSize": 12,
                "margin": 5,
              }
            }
          />
        </View>
        <View
          style={
            Object {
              "paddingHorizontal": 10,
              "width": "100%",
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "borderBottomColor": "rgba(0, 173, 238, 0.65)",
                "borderBottomWidth": 1,
                "borderColor": "#86939e",
                "flexDirection": "row",
                "height": 48,
                "transform": Array [
                  Object {
                    "translateX": 0,
                  },
                ],
              }
            }
          >
            <TextInput
              editable={true}
              onBlur={[Function]}
              onChangeText={[Function]}
              placeholder="Celular"
              placeholderTextColor="rgba(0, 173, 238, 0.65)"
              style={
                Object {
                  "color": "#00adee",
                  "flex": 1,
                  "fontSize": 16,
                  "minHeight": 40,
                }
              }
              testID="cellphoneInput"
              underlineColorAndroid="transparent"
              value=""
            />
          </View>
          <Text
            style={
              Object {
                "color": "#dc3545",
                "fontSize": 12,
                "margin": 5,
              }
            }
          />
        </View>
        <View
          style={
            Object {
              "paddingHorizontal": 10,
              "width": "100%",
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "borderBottomColor": "rgba(0, 173, 238, 0.65)",
                "borderBottomWidth": 1,
                "borderColor": "#86939e",
                "flexDirection": "row",
                "height": 48,
                "transform": Array [
                  Object {
                    "translateX": 0,
                  },
                ],
              }
            }
          >
            <TextInput
              editable={true}
              onBlur={[Function]}
              onChangeText={[Function]}
              placeholder="Dirección"
              placeholderTextColor="rgba(0, 173, 238, 0.65)"
              style={
                Object {
                  "color": "#00adee",
                  "flex": 1,
                  "fontSize": 16,
                  "minHeight": 40,
                }
              }
              testID="addressInput"
              underlineColorAndroid="transparent"
              value=""
            />
          </View>
          <Text
            style={
              Object {
                "color": "#dc3545",
                "fontSize": 12,
                "margin": 5,
              }
            }
          />
        </View>
        <View
          style={
            Object {
              "marginTop": 30,
              "paddingBottom": 15,
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "borderRadius": 3,
                  "overflow": "hidden",
                },
                Object {
                  "borderRadius": 3,
                },
                undefined,
                false,
              ]
            }
          >
            <View
              accessibilityRole="button"
              accessibilityState={
                Object {
                  "busy": false,
                  "disabled": true,
                }
              }
              accessible={true}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "opacity": 1,
                }
              }
              testID="submitButton"
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "hsl(208, 8%, 90%)",
                    "borderColor": "#2089dc",
                    "borderRadius": 20,
                    "borderWidth": 0,
                    "flexDirection": "row",
                    "height": 40,
                    "justifyContent": "center",
                    "opacity": 0.65,
                    "padding": 8,
                  }
                }
              >
                <Text
                  style={
                    Object {
                      "color": "hsl(208, 8%, 63%)",
                      "fontSize": 18,
                      "paddingVertical": 1,
                      "textAlign": "center",
                    }
                  }
                >
                  CREAR
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
